File: src-ts/browser/playwright.ts
/**
 * Browser backend using Playwright (native Node.js).
 *
 * Install: ``npm install playwright && npx playwright install chromium``.
 * Set ``browser.backend: playwright`` in config.
 */
import type { BrowserBackend, ElementRect, PageInfo, ScrollOptions, Viewport } from "./index.js";

function log(msg: string): void {
  process.stderr.write(`[spark-e2e] ${msg}\n`);
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type PlaywrightAny = any;

export class PlaywrightBackend implements BrowserBackend {
  private pw: PlaywrightAny = null;
  private browser: PlaywrightAny = null;
  private page: PlaywrightAny = null;

  private async ensureBrowser(): Promise<PlaywrightAny> {
    if (this.page) return this.page;

    const mod = await this.loadPlaywright();
    log("Starting Playwright browser (Chromium, headless)");
    this.pw = { stop: async () => { await mod.chromium.stop?.(); } };
    this.browser = await mod.chromium.launch({ headless: true });
    this.page = await this.browser.newPage();
    return this.page;
  }

  /**
   * Resolve playwright from multiple locations:
   * 1. Project CWD (for npx / global spark-e2e + local playwright)
   * 2. Global npm prefix (for global spark-e2e + global playwright)
   * 3. Bare specifier (for spark-e2e installed as project dependency)
   */
  private async loadPlaywright(): Promise<PlaywrightAny> {
    const { createRequire } = await import("node:module");
    const { join } = await import("node:path");

    const candidates: string[] = [];

    // 1. Project CWD — covers npx spark-e2e with project-local playwright
    try {
      const req = createRequire(join(process.cwd(), "package.json"));
      candidates.push(req.resolve("playwright"));
    } catch { /* not in project */ }

    // 2. Global npm — covers global spark-e2e + global playwright
    try {
      const { execSync } = await import("node:child_process");
      const globalRoot = execSync("npm root -g", { encoding: "utf-8", timeout: 5000 }).trim();
      if (globalRoot) {
        const req = createRequire(join(globalRoot, "package.json"));
        candidates.push(req.resolve("playwright"));
      }
    } catch { /* no global npm or no playwright there */ }

    // 3. Bare specifier — covers spark-e2e installed as project dep
    candidates.push("playwright");

    const errors: string[] = [];
    for (const c of candidates) {
      try {
        return await import(c);
      } catch (e) {
        errors.push(`${c}: ${(e as Error).message}`);
      }
    }

    throw new Error(
      `Playwright not found. Install it:\n` +
      `  npm install playwright && npx playwright install chromium\n` +
      `Resolution attempts:\n${errors.map(e => "  - " + e).join("\n")}`
    );
  }

  async captureScreenshot(opts?: {
    viewport?: Viewport;
    reload?: boolean;
    delay?: number;
    maxDim?: number;
    fullPage?: boolean;
  }): Promise<Buffer> {
    const page = await this.ensureBrowser();

    if (opts?.viewport) {
      await page.setViewportSize({ width: opts.viewport.width, height: opts.viewport.height });
    }

    // reload defaults to true unless explicitly disabled with --no-reload
    if (opts?.reload !== false) {
      await page.reload();
      await page.waitForLoadState("networkidle");
    }
    // delay is independent of reload — always applied when set (seconds → ms)
    if ((opts?.delay ?? 0) > 0) {
      await new Promise((r) => setTimeout(r, opts!.delay! * 1000));
    }

    return page.screenshot({ type: "png", fullPage: opts?.fullPage ?? false }) as Promise<Buffer>;
  }

  async executeJs(script: string): Promise<unknown> {
    const page = await this.ensureBrowser();
    return page.evaluate(`(${script})`);
  }

  async navigate(url: string): Promise<void> {
    const page = await this.ensureBrowser();
    log(`Navigating to ${url}`);
    await page.goto(url, { waitUntil: "networkidle" });
  }

  async getPageInfo(): Promise<PageInfo> {
    const page = await this.ensureBrowser();
    return page.evaluate(() => ({
      url: window.location.href,
      title: document.title,
      width: window.innerWidth,
      height: window.innerHeight,
      scroll_x: window.scrollX,
      scroll_y: window.scrollY,
    })) as Promise<PageInfo>;
  }

  async scroll(opts?: ScrollOptions): Promise<PageInfo> {
    const page = await this.ensureBrowser();

    if (opts?.selector) {
      await page.evaluate((sel: string) => {
        const el = document.querySelector(sel);
        if (el) el.scrollIntoView({ behavior: "instant", block: "nearest" });
      }, opts.selector);
    } else {
      const x = opts?.x ?? 0;
      const y = opts?.y ?? 0;
      await page.evaluate(
        ({ px, py }: { px: number; py: number }) =>
          window.scrollTo({ top: py, left: px, behavior: "instant" } as ScrollToOptions),
        { px: x, py: y }
      );
    }

    return this.getPageInfo();
  }

  async getElementRect(selector: string): Promise<ElementRect | null> {
    const page = await this.ensureBrowser();
    try {
      const box = await page.locator(selector).boundingBox();
      if (!box) return null;
      return {
        x: box.x,
        y: box.y,
        width: box.width,
        height: box.height,
        top: box.y,
        right: box.x + box.width,
        bottom: box.y + box.height,
        left: box.x,
      };
    } catch {
      return null;
    }
  }

  async waitForSelector(selector: string, timeoutMs = 10000): Promise<void> {
    const page = await this.ensureBrowser();
    await page.waitForSelector(selector, { timeout: timeoutMs });
  }

  async waitForTimeout(ms: number): Promise<void> {
    await new Promise((r) => setTimeout(r, ms));
  }

  toDataUrl(bytes: Buffer): string {
    return "data:image/png;base64," + bytes.toString("base64");
  }

  async close(): Promise<void> {
    if (this.browser) await this.browser.close();
    if (this.pw) await this.pw.stop();
  }
}

File: src/spark_e2e/browser/playwright_.py
"""Browser backend using Playwright.

This is a Phase 2 implementation.  Install with ``pip install spark-e2e[playwright]``
and set ``browser.backend: playwright`` in your config.
"""

from __future__ import annotations

import sys
from base64 import b64encode

from .base import BrowserBackend


def _log(msg: str) -> None:
    print(f"[spark-e2e] {msg}", file=sys.stderr, flush=True)


class PlaywrightBackend(BrowserBackend):
    """Browser automation via Playwright's sync API.

    Requires ``playwright`` to be installed and browsers to be downloaded:
    ``pip install playwright && playwright install chromium``.
    """

    def __init__(self) -> None:
        self._browser = None
        self._page = None

    def _ensure_browser(self) -> None:
        """Lazy-init the Playwright browser and page."""
        if self._page is not None:
            return

        try:
            from playwright.sync_api import sync_playwright
        except ImportError:
            raise ImportError(
                "Playwright is not installed. "
                "Install it with: pip install spark-e2e[playwright]"
            )

        _log("Starting Playwright browser (Chromium, headless)")
        self._pw = sync_playwright().start()
        self._browser = self._pw.chromium.launch(headless=True)
        self._page = self._browser.new_page()

    # ── BrowserBackend interface ─────────────────────────────────────

    def capture_screenshot(
        self,
        viewport: dict | None = None,
        reload: bool = True,
        delay: float = 0.5,
        max_dim: int = 1800,
        full_page: bool = False,
    ) -> bytes:
        self._ensure_browser()

        if viewport:
            self._page.set_viewport_size({
                "width": viewport["width"],
                "height": viewport["height"],
            })

        # reload defaults to True — SPA rendering often needs a refresh
        if reload:
            self._page.reload()
            self._page.wait_for_load_state("networkidle")
        # delay is independent of reload
        if delay > 0:
            import time
            time.sleep(delay)

        png_bytes = self._page.screenshot(type="png", full_page=full_page)

        # Scale down if needed (simple PIL-free approach: just return as-is for now)
        return png_bytes

    def execute_js(self, script: str) -> object:
        self._ensure_browser()
        return self._page.evaluate(script)

    def navigate(self, url: str) -> None:
        self._ensure_browser()
        _log(f"Navigating to {url}")
        self._page.goto(url, wait_until="networkidle")

    def get_page_info(self) -> dict:
        self._ensure_browser()
        return self._page.evaluate("""() => ({
            url: window.location.href,
            title: document.title,
            width: window.innerWidth,
            height: window.innerHeight,
            scroll_x: window.scrollX,
            scroll_y: window.scrollY,
        })""")

    def scroll(
        self,
        x: int | None = None,
        y: int | None = None,
        selector: str | None = None,
    ) -> dict:
        """Scroll the page and return updated page info."""
        self._ensure_browser()

        if selector:
            self._page.evaluate(
                """(sel) => {
                    const el = document.querySelector(sel);
                    if (el) el.scrollIntoView({behavior: 'instant', block: 'nearest'});
                }""",
                selector,
            )
        else:
            _x = x or 0
            _y = y or 0
            self._page.evaluate(
                f"window.scrollTo({{top: {_y}, left: {_x}, behavior: 'instant'}});"
            )

        return self.get_page_info()

    def get_element_rect(self, selector: str) -> dict | None:
        self._ensure_browser()
        try:
            box = self._page.locator(selector).bounding_box()
            if box is None:
                return None
            return {
                "x": box["x"],
                "y": box["y"],
                "width": box["width"],
                "height": box["height"],
                "top": box["y"],
                "right": box["x"] + box["width"],
                "bottom": box["y"] + box["height"],
                "left": box["x"],
            }
        except Exception:
            return None

    def wait_for_selector(self, selector: str, timeout_ms: int = 10000) -> None:
        """Wait for a CSS selector to appear in the DOM."""
        self._ensure_browser()
        self._page.wait_for_selector(selector, timeout=timeout_ms)

    def wait_for_timeout(self, ms: int) -> None:
        """Pause for a fixed duration."""
        import time
        time.sleep(ms / 1000)

    def to_data_url(self, image_bytes: bytes) -> str:
        """Convert image bytes to base64 data URL."""
        return "data:image/png;base64," + b64encode(image_bytes).decode("ascii")

    def close(self) -> None:
        if self._browser:
            self._browser.close()
        if hasattr(self, "_pw") and self._pw:
            self._pw.stop()
